diags/diag_nvme.c

```diff
@@ -17,8 +17,102 @@
  * USA.
  */
 
+#include <getopt.h>
+#include <regex.h>
+#include <servicelog-1/servicelog.h>
+#include <stdlib.h>
+#include <sys/utsname.h>
 #include "diag_nvme.h"
 
+int diagnose_nvme(char *device_name);
+int log_event(uint64_t *event_id, struct nvme_ibm_vpd *vpd, char *controller_name, uint8_t severity, char *description, unsigned char *raw_data, uint32_t raw_data_len);
+void print_usage(char *command);
+int regex_controller(char *controller_name, char *device_name);
+
+int main(int argc, char *argv[]) {
+	int opt, rc = 0;
+
+	static struct option long_options[] = {
+		{"help", no_argument, NULL, 'h'},
+		{0, 0, 0, 0}
+	};
+
+	while ((opt = getopt_long(argc, argv, "h", long_options, NULL)) != -1) {
+		switch (opt) {
+		case 'h':
+			print_usage(argv[0]);
+			return 0;
+			break;
+		case '?':
+			print_usage(argv[0]);
+			return -1;
+		default:
+			print_usage(argv[0]);
+			return -1;
+		}
+	}
+
+	while (optind < argc) {
+		rc += diagnose_nvme(argv[optind]);
+		fprintf(stdout, "\n");
+		optind++;
+	}
+
+	return rc;
+}
+
+/* diagnose_nvme - Diagnose NVMe device health status
+ * @device_name - Name of the device to be diagnosed, expected format is nvme[0-9]+
+ *
+ * Diagnose will check SMART data retrieved and identify any failures that need to be reported in
+ * the servicelog database.
+ *
+ * Return - Return 0 on success, negative number on failure
+ */
+int diagnose_nvme(char *device_name) {
+	char dev_path[PATH_MAX], controller_name[NAME_MAX] = { '\0' }, description[DESCR_LENGTH];
+	int fd, rc = 0, tmp_rc, time = 0, interval = 5;
+	struct nvme_smart_log_page smart_log = { 0 };
+	struct nvme_ibm_vpd vpd = { 0 };
+        uint64_t event_id;
+	uint8_t severity;
+
+	tmp_rc = regex_controller(controller_name, device_name);
+	if (tmp_rc != 0)
+		return -1;
+	snprintf(dev_path,sizeof(dev_path), "/dev/%s", device_name);
+	fd = open_nvme(dev_path);
+	if (fd < 0)
+		return -1;
+	fprintf(stdout, "Running diagnostics for %s\n", device_name);
+
+	/* After system boot VPD data might not be ready yet, so wait until VPD log page is ready */
+	while ((tmp_rc = get_ibm_vpd_log_page(fd, &vpd)) == 0x0093 && time <= MAX_TIME) {
```

Comment:
What does 0x0093 refer to? Can we consider using #define to define some of these values, instead of using constants?

Reply:
Sure, I can do that, I actually need to correct the value here, the proper value is supposed to be 0x00F3 to indicate that VPD data is not ready yet and we need to retry the operation later.

Reply:
Change done, thank you for the feedback.